File: Emisor/hamming.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <string>
#include <bitset>
#include <algorithm>
#include <cstdlib>
#include <ctime>
#include <winsock2.h>
#include <chrono>
#include <fstream>

using namespace std;
using namespace std::chrono;

// Función que calcula si un número es potencia de dos
bool isPowerOfTwo(int n) {
    return (n != 0) && ((n & (n - 1)) == 0);
}

// Función para convertir una cadena a binario
string stringToBinary(const string& input) {
    string binaryString;

    for (size_t i = 0; i < input.size() + 1; ++i) {
        char c = input[i];
        bitset<8> binary(c);
        binaryString += binary.to_string();
    }
    
    return binaryString;
}

pair<string, pair<bool, int>> hamming(string message) {
    int n, m, r = 0;
    string binaryMessage, finalMessage;
    bool flag = false, noisy = false;
    int noiseCount = 0;

    // Inicializa la semilla del generador de números aleatorios
    srand(static_cast<unsigned int>(time(0)));

    cout << "--------------- Codigo de Hamming ---------------" << endl << endl;

    while (!flag) {
        if (message.length() > 0) { // Si no es una cadena vacia
            flag = true;
        } else {
            cout << "La longitud de la cadena debe ser mayor a 0, intente nuevamente." << endl;
        }
    }

    binaryMessage = stringToBinary(message); // Convierte la cadena a binario
    // Eliminar los espacios en blanco al final de binaryMessage
    binaryMessage.erase(find_if(binaryMessage.rbegin(), binaryMessage.rend(), [](unsigned char ch) {
        return !isspace(ch);
    }).base(), binaryMessage.end());

    m = binaryMessage.length();

    while (pow(2, r) < (m + r + 1)) { // Calcula el número de bits de paridad
        r++;
    }
    
    n = m + r;

    vector<char> hammingCode(n, '0'); // Arreglo para el mensaje codificado

    int j = 0;
    for (int i = 1; i <= n; i++) {
        if (isPowerOfTwo(i)) { // Si es potencia de dos
            hammingCode[i - 1] = 'p'; // Se inserta un caracter 'p' para marcar el espacio
        } else {
            hammingCode[i - 1] = binaryMessage[j];
            j++;
        }
    }

    // Calculo de los bits de paridad
    for (int i = 1; i <= n; i++) {
        if (isPowerOfTwo(i)) {
            int conteo = 0;

            for (int j = i; j <= n; j++) {
                if ((j & i) != 0) {
                    if (hammingCode[j - 1] == '1') {
                        conteo++;
                    }
                }
            }

            if (conteo % 2 == 0) {
                hammingCode[i - 1] = '0';
            } else {
                hammingCode[i - 1] = '1';
            }
        }
    }

    cout << endl << "--------------- Cadena con ruido ---------------" << endl;

    // Introduciendo ruido aleatoriamente en la cadena
    for (char& c : hammingCode) {
        int randomNumber = rand() % 100;
        
        if (randomNumber == 0) { // Si hay probabilidad de 1/100
            c = (c == '0') ? '1' : '0'; // Cambia el valor

            cout << "\033[31m" << c << "\033[0m"; // Color rojo cambio de bit
            noisy = true;
            noiseCount++;
        } else {
            cout << c; // Color negro bit normal
        }

        finalMessage += c;
    }

    finalMessage += "END";

    return make_pair(finalMessage, make_pair(noisy, noiseCount));
}

int main() {
    // Initialize Winsock
    WSADATA wsaData;
    WSAStartup(MAKEWORD(2, 2), &wsaData);

    // Create a socket
    SOCKET s = socket(AF_INET, SOCK_STREAM, 0);
    if (s == INVALID_SOCKET) {
        cerr << "Could not create socket: " << WSAGetLastError() << endl;
        return 1;
    }

    // Setup the sockaddr_in structure
    sockaddr_in server;
    server.sin_family = AF_INET;
    server.sin_addr.s_addr = inet_addr("127.0.0.1");
    server.sin_port = htons(12345);

    // Connect to the server
    if (connect(s, (sockaddr*)&server, sizeof(server)) < 0) {
        cerr << "Connection failed: " << WSAGetLastError() << endl;
        return 1;
    }

    vector<double> durations;
    vector<string> noisyResults;
    vector<int> noiseCounts;

    // Generar el mensaje codificado con Hamming
    for (int i = 0; i < 100; i++) {
        string message = "Mensaje Prueba " + to_string(i);

        auto result = hamming(message);
        string hammingCode = result.first;
        bool noisy = result.second.first;
        int noiseCount = result.second.second;

        auto start = high_resolution_clock::now();

        // Send data
        if (send(s, hammingCode.c_str(), hammingCode.size(), 0) < 0) {
            cerr << "Send failed: " << WSAGetLastError() << endl;
            return 1;
        }

        auto end = high_resolution_clock::now();
        double duration = duration_cast<microseconds>(end - start).count();
        durations.push_back(duration);

        noisyResults.push_back(noisy ? "True" : "False");
        noiseCounts.push_back(noiseCount);
    }

    closesocket(s);
    WSACleanup();

    // Output the durations to a file for further analysis
    ofstream durationFile("../Statistics/Hamming/Emisor/durations.txt");
    if (!durationFile) {
        cerr << "Error opening durations file for writing" << endl;
        return 1;
    }
    for (double duration : durations) {
        durationFile << duration << endl;
    }
    durationFile.close();

    // Output the noisy results to a file
    ofstream noisyFile("../Statistics/Hamming/Emisor/noisy_results.txt");
    if (!noisyFile) {
        cerr << "Error opening noisy results file for writing" << endl;
        return 1;
    }
    for (const string& result : noisyResults) {
        noisyFile << result << endl;
    }
    noisyFile.close();

    // Output the noise counts to a file
    ofstream noiseCountFile("../Statistics/Hamming/Emisor/noise_counts.txt");
    if (!noiseCountFile) {
        cerr << "Error opening noise counts file for writing" << endl;
        return 1;
    }
    for (int count : noiseCounts) {
        noiseCountFile << count << endl;
    }
    noiseCountFile.close();

    return 0;
}

File: Receptor/hamming.py
def is_power_of_two(n):
    return (n != 0) and ((n & (n - 1)) == 0)

def binary_to_text(binary_str):
    # Divide el mensaje binario en bloques de 8 bits
    binary_list = [binary_str[i:i+8] for i in range(0, len(binary_str), 8)]
    # Convierte cada bloque de 8 bits a un carácter ASCII
    text = ''.join([chr(int(b, 2)) for b in binary_list])
    return text

def hammingCode(binary_message):
    r = 0
    flag = False

    print("--------------- Receptor en codigo de Hamming ---------------\n")

    while not flag:
        m = len(binary_message)

        if m > 0:
            while (2 ** r) < (m + r + 1):
                r += 1
            flag = True
        else:
            print("La longitud del mensaje debe ser mayor a 0, intente nuevamente.")

    n = m
    hamming_code = list(binary_message)

    # Verificación de los bits de paridad y corrección de errores
    error_position = 0
    error_count = 0

    for i in range(1, n + 1):
        if is_power_of_two(i):
            conteo = 0
            for j in range(i, n + 1):
                if (j & i) != 0:
                    if hamming_code[j - 1] == '1':
                        conteo += 1
            if conteo % 2 != 0:
                error_position += i

    if error_position > n:
        print("La cadena tiene más de dos errores y no pueden ser identificados")
        return True, "2+"

    elif error_position != 0:
        print("\n\n---------------------------------------------------------------------------")
        print(f"Error en la posición: {error_position}")
        hamming_code[error_position - 1] = '1' if hamming_code[error_position - 1] == '0' else '0'

        print("Mensaje corregido: ", end="")
        print("".join(hamming_code))

        # Extracción del mensaje original
        original_message = ""
        for i in range(1, n + 1):
            if not is_power_of_two(i):
                original_message += hamming_code[i - 1]

        print(f"\n\n--------------- Mensaje original --------------- \n{binary_to_text(original_message)}")
        return True, "1"

    elif error_position == 0:
        # Extracción del mensaje original
        original_message = ""
        for i in range(1, n + 1):
            if not is_power_of_two(i):
                original_message += hamming_code[i - 1]

        print(f"\n\n--------------- Mensaje original --------------- \n{binary_to_text(original_message)}")
        return False, "0"

def main():
    import socket
    import time

    # Configuration
    host = 'localhost'
    port = 12345

    # Create a TCP/IP socket
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.bind((host, port))
    sock.listen(1)

    print(f"Listening on {host}:{port}...")

    while True:
        connection, client_address = sock.accept()
        try:
            print(f"Connection from {client_address}")

            durations = []
            noisy = []
            error_counts = []
            numTest = 100

            for _ in range(numTest):
                startTime = time.time()

                # Receive the data
                data = b""
                while True:
                    chunk = connection.recv(4096)
                    if b"END" in chunk:
                        data += chunk[:chunk.find(b"END")]
                        break
                    data += chunk

                if data:
                    binaryMessage = data.decode()
                    response, error_count = hammingCode(binaryMessage)

                endTime = time.time()
                durations.append((endTime - startTime) * 1e6)  # Convert to microseconds
                noisy.append(response)
                error_counts.append(error_count)

                # Print results for debugging
                print(f"Processed message {_ + 1}/{numTest}, Duration: {(endTime - startTime) * 1e6} µs, Noisy: {response}, Errors: {error_count}")

            print("\n\n------------------------------------------------------------ \n")

            # Output the durations to a file for further analysis
            with open("Statistics/Hamming/Receptor/receive_durations.txt", "w") as outfile:
                for duration in durations:
                    outfile.write(f"{duration}\n")

            with open("Statistics/Hamming/Receptor/noisy_output.txt", "w") as outfile2:
                for k in noisy:
                    outfile2.write(f"{k}\n")

            with open("Statistics/Hamming/Receptor/error_counts.txt", "w") as outfile3:
                for count in error_counts:
                    outfile3.write(f"{count}\n")

        finally:
            connection.close()

if __name__ == "__main__":
    main()
